apps/auto_viz/Makefile

```diff
@@ -5,47 +5,47 @@ naive_up naive_down \
 lessnaive_up lessnaive_down \
 complex_up complex_down
 
-LIBRARIES = $(foreach V,$(VARIANTS),$(BIN)/auto_viz_demo_$(V).a)
-OUTPUTS = $(foreach V,$(VARIANTS),$(BIN)/out_$(V).png)
+LIBRARIES = $(foreach V,$(VARIANTS),$(BIN)/$(HL_TARGET)/auto_viz_demo_$(V).a)
+OUTPUTS = $(foreach V,$(VARIANTS),$(BIN)/$(HL_TARGET)/out_$(V).png)
 
 TRACE_TARGET=$(HL_TARGET)-trace_all
 
 all: $(OUTPUTS)
 
-$(BIN)/auto_viz_demo.generator: auto_viz_demo_generator.cpp $(GENERATOR_DEPS)
+$(GENERATOR_BIN)/auto_viz_demo.generator: auto_viz_demo_generator.cpp $(GENERATOR_DEPS)
 	@mkdir -p $(@D)
 	$(CXX) $(CXXFLAGS) -g $(filter-out %.h,$^) -o $@ $(LDFLAGS) $(HALIDE_SYSTEM_LIBS)
 
-$(BIN)/auto_viz_demo_%.a: $(BIN)/auto_viz_demo.generator
+$(BIN)/$(HL_TARGET)/auto_viz_demo_%.a: $(GENERATOR_BIN)/auto_viz_demo.generator
 	@mkdir -p $(@D)
-	$^ -g auto_viz_demo -o $(BIN) -f auto_viz_demo_$* \
+	$^ -g auto_viz_demo -o $(@D) -f auto_viz_demo_$* \
 	target=$(TRACE_TARGET)-no_runtime \
 	schedule_type=$$(echo $* | cut -d_ -f1) \
 	upsample=$$(echo $* | cut -d_ -f2 | sed 's/up/true/;s/down/false/')
 
-$(BIN)/runtime.a: $(BIN)/auto_viz_demo.generator
+$(BIN)/%/runtime.a: $(GENERATOR_BIN)/auto_viz_demo.generator
 	@mkdir -p $(@D)
-	$^ -r runtime -o $(BIN) target=$(TRACE_TARGET)
+	$^ -r runtime -o $(@D) target=$(TRACE_TARGET)
 
-$(BIN)/auto_viz_demo: auto_viz_demo.cpp $(LIBRARIES) $(BIN)/runtime.a
+$(BIN)/%/auto_viz_demo: auto_viz_demo.cpp $(LIBRARIES) $(BIN)/$(HL_TARGET)/runtime.a
 	@mkdir -p $(@D)
-	$(CXX) $(CXXFLAGS) -I $(BIN) $^ -o $@ $(IMAGE_IO_FLAGS) $(LDFLAGS)
+	$(CXX) $(CXXFLAGS) -I$(BIN)/$* $^ -o $@ $(IMAGE_IO_FLAGS) $(LDFLAGS)
 
-$(BIN)/out_%_up.png: $(BIN)/auto_viz_demo
+$(BIN)/$(HL_TARGET)/out_%_up.png: $(BIN)/$(HL_TARGET)/auto_viz_demo
```

Comment:
I think it would be better to use a big macro to stamp out the filter-specific rules so that we can avoid all explicit use of HL_TARGET here as you did in the later Makefiles.

Reply:
meh, ok

Reply:
done